### PR TITLE
[docs] [refactor] Add help messages for datadir path mangling

### PR DESCRIPTION
**Pull request by jamesob**
Change `-conf`'s help message to indicate that relative path values will be prefixed by the datadir path. This behavior probably merits clarification; it's kind of confusing when attempting to specify a configuration file in the current directory with `-conf=bitcoin.conf`, but instead loading the `bitcoin.conf` file in ~/.bitcoin datadir.

### Edit

This PR has been modified to document all cases where relative path configurations are modified to be under datadir. A small refactoring has also been added which consolidates this normalization.

### CODE REVIEW

**Comment by laanwj** [comment]
Not just this one, all relative files and directories specified either on the command line or through RPC are supposed to be relative to the data directory.

**Comment by MarcoFalke** [comment]
Wasn't walletdir temporarily allowed to be relative to the cwd?

**Comment by jamesob** [comment]
Sounds like we don't have any documentation indicating that all path arguments are relative to datadir;  that information should live somewhere (preferably in a single place that isn't easy to miss). One option is to put it in a paragraph between "Usage" and "Options" in `{bitcoind,bitcoin-cli,bitcoin-qt} -h`. Any other suggestions?

**Comment by MarcoFalke** [comment]
Options that take paths or file names are:

* `-rpccookiefile=<loc>`
* `-walletdir=<dir>`
* `-wallet=<file>`
* `-datadir=<dir>`
* `-conf=<file>`

So, adding a general paragraph only makes sense if it holds true for all paths. Otherwise, I'd prefer if each command line help was clarified individually.

**Comment by jamesob** [comment]
@MarcoFalke thanks for that list. I've looked at how each option on the list is parsed (as well as a few others):

- `-rpccookiefile=<loc>`: Is put under datadir if relative.
- `-walletdir=<dir>`: Isn't allowed to be relative.
- `-wallet=<file>`: Moved under walletdir with `fs::absolute(walletFile, GetWalletDir())`.
- `-datadir=<dir>`: Can be specified as relative but will warn. Obviously not put under itself.
- `-conf=<file>`: Is put under datadir if relative.
- `-loadblock=<file>`: Interpreted literally -- isn't modified if relative. Passed directly to `fsbridge::fopen`.
- `-debuglogfile=<file>`: Is put under datadir if relative.
- `-pid=<file>`: Is put under datadir if relative. 

So: each of these options is either not allowed to be relative or is put under datadir if relative with the exception of `wallet` and `loadblock`. Because we don't uniformly treat relative paths as relative to datadir, I've updated only the affected help messages.

I've also added a refactoring commit which consolidates the "normalization" of relative paths under datadir.

**Comment by laanwj** [comment]
Thanks, utACK

File: src/util.cpp [hunk 1]
```diff
@@ -936,3 +927,10 @@ int64_t GetStartupTime()
 {
     return nStartupTime;
 }
+
+fs::path NormalizePathArg(const fs::path& pathArg, bool fNetSpecific)
```
Comment (promag) [inline src/util.cpp]: I'd rename to `GetAbsolutePath` or more specific `GetAbsoluteDataPath`:
```cpp
fs::path GetAbsoluteDataPath(const fs::path& path, bool net_specific)
```
I other contexts/languages, normalization can be interpreted as transforming from `foo/../bar/` to `bar`.

Nit, snake_case arguments.

File: src/wallet/init.cpp [hunk 2]
```diff
@@ -7,6 +7,7 @@
 
 #include <net.h>
 #include <util.h>
+#include <fs.h>
```
Comment (promag) [inline src/wallet/init.cpp]: Nit, unrelated change, remove.

**Review by promag — COMMENTED** [review]
Concept ACK.

**Comment by jamesob** [comment]
Thanks for the feedback @promag. It's been incorporated and pushed.

File: src/util.cpp [hunk 3]
```diff
@@ -936,3 +927,10 @@ int64_t GetStartupTime()
 {
     return nStartupTime;
 }
+
+fs::path GetAbsolutePathForArg(const fs::path& path, bool net_specific)
+{
+    if (!path.is_complete())
+        return GetDataDir(net_specific) / path;
```
Comment (ryanofsky) [inline src/util.cpp]: Would be a little better to return: `return fs::absolute(path, GetDataDir(net_specific))` to make it clear this returns an absolute path. You could also drop the `is_complete` check above if you did this.

File: src/util.cpp [hunk 4]
```diff
@@ -936,3 +927,10 @@ int64_t GetStartupTime()
 {
     return nStartupTime;
 }
+
+fs::path GetAbsolutePathForArg(const fs::path& path, bool net_specific)
+{
+    if (!path.is_complete())
```
Comment (ryanofsky) [inline src/util.cpp]: Could eliminate this check, but if you will keep it [developer notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md) suggest adding braces or putting if body on same line as condition.

File: src/util.h [hunk 5]
```diff
@@ -191,6 +191,16 @@ bool OpenDebugLog();
...
 
+/**
+ * Most paths passed as configuration arguments are treated as relative to
+ * the datadir if they are not absolute.
+ *
+ * @param path The path to be conditionally prefixed with datadir.
+ * @param net_specific Forwarded to GetDataDir().
+ * @return The normalized path.
+ */
+fs::path GetAbsolutePathForArg(const fs::path& path, bool net_specific = true);
```
Comment (ryanofsky) [inline src/util.h]: "ForArg" in function name seems a little redundant. Just a suggestion, but I might call it something `GetDataPath` or `GetDataDirPath`.

**Review by ryanofsky — COMMENTED** [review]
utACK 23d295aa9e16cb3591bbd8bb0f5ce026b861b7b6

File: src/util.cpp (hunk 3 above)
Comment (jamesob) [reply src/util.cpp]: Great, thanks.

File: src/util.h (hunk 5 above)
Comment (jamesob) [reply src/util.h]: My only reservation here is that the path won't necessarily be in datadir, so a name with Data or DataDir seems confusing. "Arg" here is meant to correspond to an argument to the binary, not the function, but maybe that isn't clear.

File: src/util.cpp (hunk 4 above)
Comment (jamesob) [reply src/util.cpp]: Taking your advice below, but the reminder about convention is appreciated.

File: src/util.h (hunk 5 above)
Comment (jamesob) [reply src/util.h]: I'll try `AbsPathForConfigVal`.

**Comment by jamesob** [comment]
Thanks for the good feedback, @ryanofsky. Incorporated and pushed.

**Comment by ajtowns** [comment]
Could update the commit title to match the renamed function :)

**Comment by jamesob** [comment]
d'oh. Thanks, @ajtowns. Pushed.

**Comment by jnewbery** [comment]
Looks good.

Minor note that `-conf` is relative to the base datadir, but `-debuglogfile`, `-pid` and `-rpccookiefile` are relative the net-specific subdirectories. It might be better to specify that in the help text.

**Comment by sipa** [comment]
utACK 95963c120c48ce750fdb2298f3eb70034e578e8d, but needs rebase.

**Comment by jamesob** [comment]
Rebased and incorporated @jnewbery's feedback.

File: src/util.h [hunk 6]
```diff
@@ -191,6 +191,16 @@ bool OpenDebugLog();
 void ShrinkDebugFile();
 void runCommand(const std::string& strCommand);
 
+/**
+ * Most paths passed as configuration arguments are treated as relative to
+ * the datadir if they are not absolute.
+ *
+ * @param path The path to be conditionally prefixed with datadir.
+ * @param net_specific Forwarded to GetDataDir().
+ * @return The normalized path.
```
Comment (promag) [inline src/util.h]: The absolute path?

**Comment by jnewbery** [comment]
utACK 54604600c3de6cb18540c0911127173f68ad246c

**Comment by MarcoFalke** [comment]
Post-merge utACK 54604600c3de6cb18540c0911127173f68ad246c
